warehouse/oidc/models/gitlab.py

```diff
@@ -0,0 +1,288 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any
+
+from sqlalchemy import ForeignKey, String, UniqueConstraint
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.orm import Query, mapped_column
+from sqlalchemy.sql.expression import func, literal
+
+from warehouse.oidc.errors import InvalidPublisherError
+from warehouse.oidc.interfaces import SignedClaims
+from warehouse.oidc.models._core import (
+    CheckClaimCallable,
+    OIDCPublisher,
+    PendingOIDCPublisher,
+    check_claim_binary,
+)
+
+
+def _check_ci_config_ref_uri(ground_truth, signed_claim, all_signed_claims):
+    # We expect a string formatted as follows:
+    #   gitlab.com/OWNER/REPO//WORKFLOW_PATH/WORKFLOW_FILE.yml@REF
+    # where REF is the value of the `ref_path` claim.
+
+    # Defensive: GitLab should never give us an empty ci_config_ref_uri,
+    # but we check for one anyway just in case.
+    if not signed_claim:
+        raise InvalidPublisherError("The ci_config_ref_uri claim is empty")
+
+    # We need at least one of these to be non-empty
+    ref = all_signed_claims.get("ref_path")
+    sha = all_signed_claims.get("sha")
+    if not (ref or sha):
+        raise InvalidPublisherError("The ref_path and sha claims are empty")
+
+    expected = {f"{ground_truth}@{_ref}" for _ref in [ref, sha] if _ref}
```

Comment:
```suggestion
    ref_path = all_signed_claims.get("ref_path")
    sha = all_signed_claims.get("sha")
    if not (ref_path or sha):
        raise InvalidPublisherError("The ref_path and sha claims are empty")

    expected = {f"{ground_truth}@{_ref}" for _ref in [ref_path, sha] if _ref}
```
A bit of a nit, but since there's also a `ref` claim, let's be more clear here.

Reply:
fixed